src/Sylius/Bundle/CoreBundle/spec/Provider/CatalogPromotionVariantsProviderSpec.php

```diff
@@ -15,45 +15,37 @@
 
 use Doctrine\Common\Collections\ArrayCollection;
 use PhpSpec\ObjectBehavior;
+use Sylius\Bundle\CoreBundle\Provider\VariantsProviderInterface;
 use Sylius\Component\Core\Model\CatalogPromotionInterface;
-use Sylius\Component\Core\Model\ProductInterface;
 use Sylius\Component\Core\Model\ProductVariantInterface;
-use Sylius\Component\Core\Model\TaxonInterface;
 use Sylius\Component\Core\Provider\CatalogPromotionVariantsProviderInterface;
-use Sylius\Component\Core\Repository\ProductRepositoryInterface;
-use Sylius\Component\Core\Repository\ProductVariantRepositoryInterface;
-use Sylius\Component\Promotion\Model\CatalogPromotionRuleInterface;
-use Sylius\Component\Taxonomy\Repository\TaxonRepositoryInterface;
+use Sylius\Component\Core\Model\CatalogPromotionRuleInterface;
 
 final class CatalogPromotionVariantsProviderSpec extends ObjectBehavior
 {
-    function let(
-        ProductVariantRepositoryInterface $productVariantRepository
-    ): void {
-        $this->beConstructedWith($productVariantRepository);
+    function let(VariantsProviderInterface $firstProvider, VariantsProviderInterface $secondProvider): void
+    {
+        $this->beConstructedWith([$firstProvider, $secondProvider]);
     }
 
     function it_implements_catalog_promotion_products_provider_interface(): void
     {
         $this->shouldImplement(CatalogPromotionVariantsProviderInterface::class);
     }
 
-    function it_provides_products_with_configured_variants(
-        ProductVariantRepositoryInterface $productVariantRepository,
+    function it_provides_variants_configured_in_catalog_promotion_rule(
         CatalogPromotionInterface $catalogPromotion,
         CatalogPromotionRuleInterface $rule,
+        VariantsProviderInterface $firstProvider,
         ProductVariantInterface $firstVariant,
         ProductVariantInterface $secondVariant
     ): void {
         $catalogPromotion->getRules()->willReturn(new ArrayCollection([$rule->getWrappedObject()]));
         $rule->getConfiguration()->willReturn(['variants' => ['PHP_T_SHIRT_XS_WHITE', 'PHP_T_SHIRT_XS_BLACK', 'PHP_MUG']]);
+        $rule->getType()->willReturn(CatalogPromotionRuleInterface::TYPE_FOR_VARIANTS);
 
-        $productVariantRepository->findOneBy(['code' => 'PHP_T_SHIRT_XS_WHITE'])->willReturn($firstVariant);
-        $productVariantRepository->findOneBy(['code' => 'PHP_T_SHIRT_XS_BLACK'])->willReturn($secondVariant);
-        $productVariantRepository->findOneBy(['code' => 'PHP_MUG'])->willReturn(null);
-
-        $firstVariant->getCode()->willReturn('PHP_T_SHIRT_XS_WHITE');
-        $secondVariant->getCode()->willReturn('PHP_T_SHIRT_XS_BLACK');
+        $firstProvider->supports($rule)->willReturn(true);
```

Comment:
If we return `false` here and `true` in the second one, we would also check if the if statement is done properly 💃